Object storage operations

snazy opened the pull request:
API and implementations to perform long-running operations against object stores, mostly to purge files.

snazy committed:
Object storage operations (proposal)

API and implementations to perform long-running operations against object stores, mostly to purge files.

adam-christian-software commented on storage/files/README.md:
```diff
@@ -0,0 +1,70 @@
+
+# Polaris object store operations
+
+API and implementations to perform long-running operations against object stores, mostly to purge files.
```
Nit: I wouldn't limit to purging files. I could see other object store operations which might be helpful in the future.

So, I'd remove "mostly to purge files."

adam-christian-software commented on storage/files/README.md:
```diff
@@ -0,0 +1,70 @@
+implementations. 
+
+The actual object store interaction of the current implementation is delegated to Iceberg `FileIO` implementations.
+Only `FileIO` implementations that support prefix-operations (`SupportsPrefixOperations` interface) and
```
What sort of FileIO implementations are not supported then? From what I'm seeing OSSFileIO, InMemoryFileIO, & EcsFileIO?

adam-christian-software commented on storage/files/api/src/main/java/org/apache/polaris/storage/files/api/FileOperations.java:
```diff
@@ -0,0 +1,170 @@
+  /**
+   * Identifies all files referenced by the given table-metadata.
+   *
+   * <p>In case "container" files, like the metadata, manifest-list or manifest files, are not
```
When would these files not be readable? Did you have cases in mind when that would happen?

adam-christian-software commented on storage/files/api/src/main/java/org/apache/polaris/storage/files/api/FileOperations.java:
```diff
@@ -0,0 +1,170 @@
+   *
+   * @param tableMetadataLocation Iceberg table-metadata location
+   * @param deduplicate if true, attempt to deduplicate files by their location, adding additional
+   *     heap pressure to the operation. Implementations may ignore this parameter or may not
```
Is there a reason to have the implementation ignore this parameter? Like, it seems as if it would be helpful because, if it is ignored sometimes, the caller will always have to assume that it is not deduplicated. So, the caller might be doing unnecessary work.

adam-christian-software commented on storage/files/api/src/main/java/org/apache/polaris/storage/files/api/FileOperations.java:
```diff
@@ -0,0 +1,170 @@
+ * <p>All functions of this interface rather yield incomplete results and continue over throwing
+ * exceptions.
+ */
+public interface FileOperations {
```
Nit: For readability, I would introduce a FilePurgeOperations interface which could contain all of the purge operations and have this interface extend that one.

adam-christian-software commented on storage/files/api/src/main/java/org/apache/polaris/storage/files/api/PurgeSpec.java:
```diff
@@ -0,0 +1,99 @@
+   * Callback being invoked right before a file location is being submitted to be purged.
+   *
+   * <p>Due to API constraints of {@link
+   * org.apache.iceberg.io.SupportsBulkOperations#deleteFiles(Iterable)} it's barely possible to
```
Nit: I'd rephrase to "This callback allows users to identify files that failed a deletion."

adam-christian-software reviewed (commented):
This looks like a fantastic PR! Thanks for pushing forward with this. I had some smaller questions, but this looks great.

flyrain reviewed (commented):
Thanks for the PR! This seems to overlap with the existing `TableCleanupTaskHandler`. My understanding is that the direction is to invest in a more general task execution framework (e.g., a delegation service outside of Polaris), rather than re-implementing file-level operations in individual features. It would be great to align this work with that direction.

snazy commented on storage/files/impl/src/main/java/org/apache/polaris/storage/files/impl/FileOperationsImpl.java:
```diff
@@ -0,0 +1,471 @@
+
+  @Override
+  public Stream<FileSpec> identifyIcebergViewFiles(
+      @Nonnull String viewMetadataLocation, boolean deduplicate) {
```
Not sure a TODO is reasonable (it gives the impression that the code as it is now needs some change).

snazy commented on storage/files/api/src/main/java/org/apache/polaris/storage/files/api/PurgeStats.java:
```diff
@@ -0,0 +1,42 @@
+  /**
+   * The number of purged files.
+   *
+   * <p>The returned value may be wrong and include non-existing files.
```
I'd prefer to leave it ;)
The reason is that this object-storage API doesn't _require_ us to use Iceberg's `FileIO`. We could access the object-store SDKs directly, which would be more efficient and yield correct numbers including deletion-error information per file/object. But that would be some more work.

adam-christian-software replied:
AH OK. Sounds good. I'm good to resolve.

flyrain commented on storage/files/impl/src/testFixtures/java/org/apache/polaris/storage/files/impl/IcebergFixtures.java:
```diff
@@ -0,0 +1,194 @@
+import org.projectnessie.catalog.formats.iceberg.manifest.IcebergManifestListWriter;
+import org.projectnessie.catalog.formats.iceberg.manifest.IcebergManifestListWriterSpec;
+import org.projectnessie.catalog.formats.iceberg.meta.IcebergPartitionSpec;
+import org.projectnessie.catalog.formats.iceberg.meta.IcebergSchema;
```
I do not think Polaris should depend on the Nessie project for Iceberg metadata. What if Iceberg itself changed, while Nessie doesn't change them or haven't change yet? cc @dennishuo @singhpk234 @adam-christian-software

snazy replied:
These classes are only needed for simpler tests, which are not possible or only possible with a lot of hacks with Iceberg Java code. These classes would only break, if the Iceberg _spec_ changes in a breaking way.

flyrain replied:
I do not fully agree with this.

These classes are not just test helpers. They introduce a parallel Iceberg metadata and behavior model, even if used only in tests today. That means they still encode assumptions about Iceberg semantics, which increases long term maintenance risk.

The idea that they would only break with a breaking Iceberg spec change is also optimistic. Iceberg often evolves through additive or clarifying changes that do not break the spec but can still invalidate assumptions in helper classes like these.

Finally, this creates an indirect dependency where Polaris has to wait for Nessie to catch up whenever Iceberg changes. That coupling feels unnecessary and makes it harder for Polaris to stay closely aligned with upstream Iceberg behavior.

adam-christian-software replied:
@pingtimeout, @snazy, & @flyrain - I'm fine merging this in as-is with a quick follow up to remove the Nessie dependency. I played around with it and I have a version of IcebergFixtures that can rely directly on Iceberg without Project Nessie (see below).

```
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.polaris.storage.files.impl;

import static java.lang.String.format;
import static java.nio.charset.StandardCharsets.UTF_8;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Map;
import org.apache.iceberg.DataFile;
import org.apache.iceberg.DataFiles;
import org.apache.iceberg.FileFormat;
import org.apache.iceberg.IcebergBridge;
import org.apache.iceberg.ManifestFile;
import org.apache.iceberg.ManifestWriter;
import org.apache.iceberg.PartitionSpec;
import org.apache.iceberg.Schema;
import org.apache.iceberg.TableMetadata;
import org.apache.iceberg.TableMetadataParser;
import org.apache.iceberg.avro.Avro;
import org.apache.iceberg.io.FileAppender;
import org.apache.iceberg.io.InputFile;
import org.apache.iceberg.io.OutputFile;
import org.apache.iceberg.io.PositionOutputStream;
import org.apache.iceberg.types.Types;

public class IcebergFixtures {
  public final Schema schema;
  public final PartitionSpec spec;
  public final TableMetadata tableMetadata;
  public final String tableMetadataString;
  public final byte[] tableMetadataBytes;

  public final String prefix;
  public final int numSnapshots;
  public final int numManifestFiles;
  public final int numDataFiles;

  public IcebergFixtures(String prefix, int numSnapshots, int numManifestFiles, int numDataFiles) {
    this.prefix = prefix;
    this.numSnapshots = numSnapshots;
    this.numManifestFiles = numManifestFiles;
    this.numDataFiles = numDataFiles;

    schema = new Schema(1, Types.NestedField.required(1, "foo", Types.StringType.get()));
    spec = PartitionSpec.unpartitioned();

    var tableMetadataBuilder =
        TableMetadata.buildFrom(
            TableMetadata.newTableMetadata(schema, spec, prefix, Map.of()).withUUID());
    for (var snapshotId = 1; snapshotId <= numSnapshots; snapshotId++) {
      var manifestList = manifestListPath(snapshotId);
      var snapshot =
          IcebergBridge.mockSnapshot(
              snapshotId + 1,
              snapshotId + 1,
              snapshotId > 0 ? (long) snapshotId : null,
              System.currentTimeMillis(),
              "APPEND",
              Map.of(),
              schema.schemaId(),
              manifestList,
              (long) numManifestFiles * numManifestFiles);
      tableMetadataBuilder.addSnapshot(snapshot);
    }
    tableMetadata = tableMetadataBuilder.build();

    tableMetadataString = TableMetadataParser.toJson(tableMetadata);
    tableMetadataBytes = tableMetadataString.getBytes(UTF_8);
  }

  public String manifestListPath(int snapshotId) {
    return format("%s%05d/snap-%d.avro", prefix, snapshotId, snapshotId);
  }

  public byte[] serializedManifestList(long snapshotId) {
    var output = new ByteArrayOutputStream();
    OutputFile outputFile = new InMemoryOutputFile(output);

    try (FileAppender<ManifestFile> manifestListWriter =
        Avro.write(outputFile)
            .schema(ManifestFile.schema())
            .named("manifest_file")
            .overwrite()
            .build()) {
      for (int i = 0; i < numManifestFiles; i++) {
        var manifestPath = manifestFilePath(snapshotId, i);

        // Create a manifest file and get its ManifestFile metadata
        var tempManifestOutput = new ByteArrayOutputStream();
        OutputFile tempManifestOutputFile =
            new InMemoryOutputFile(tempManifestOutput, manifestPath);
        ManifestWriter<DataFile> tempWriter =
            org.apache.iceberg.ManifestFiles.write(2, spec, tempManifestOutputFile, snapshotId);
        for (int j = 0; j < numDataFiles; j++) {
          var dataFilePath = format("%s%05d/%05d/%05d/data.parquet", prefix, snapshotId, i, j);
          DataFile dataFile =
              DataFiles.builder(spec)
                  .withFormat(FileFormat.PARQUET)
                  .withPath(dataFilePath)
                  .withFileSizeInBytes(1024)
                  .withRecordCount(1)
                  .build();
          tempWriter.add(dataFile);
        }
        tempWriter.close();
        ManifestFile manifestFile = tempWriter.toManifestFile();

        manifestListWriter.add(manifestFile);
      }
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return output.toByteArray();
  }

  public String manifestFilePath(long snapshotId, int file) {
    return format("%s%05d/%05d/xyz-m-manifest.avro", prefix, snapshotId, file);
  }

  public byte[] serializedManifestFile(long snapshotId, int manifest, String path) {
    var output = new ByteArrayOutputStream();
    OutputFile outputFile = new InMemoryOutputFile(output);

    try (ManifestWriter<DataFile> manifestFileWriter =
        org.apache.iceberg.ManifestFiles.write(2, spec, outputFile, snapshotId)) {
      for (int i = 0; i < numDataFiles; i++) {
        var dataFilePath = format("%s%05d/%05d/%05d/data.parquet", prefix, snapshotId, manifest, i);
        DataFile dataFile =
            DataFiles.builder(spec)
                .withFormat(FileFormat.PARQUET)
                .withPath(dataFilePath)
                .withFileSizeInBytes(1024)
                .withRecordCount(1)
                .build();
        manifestFileWriter.add(dataFile);
      }
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
    return output.toByteArray();
  }

  private record InMemoryOutputFile(ByteArrayOutputStream output, String path) implements OutputFile {
      InMemoryOutputFile(ByteArrayOutputStream output) {
        this(output, "in-memory");
      }

    @Override
      public PositionOutputStream create() {
        return new ByteArrayPositionOutputStream(output);
      }

      @Override
      public PositionOutputStream createOrOverwrite() {
        output.reset();
        return new ByteArrayPositionOutputStream(output);
      }

      @Override
      public String location() {
        return path;
      }

      @Override
      public InputFile toInputFile() {
        throw new UnsupportedOperationException("Not implemented");
      }
    }

  private static class ByteArrayPositionOutputStream
      extends org.apache.iceberg.io.PositionOutputStream {
    private final ByteArrayOutputStream output;

    ByteArrayPositionOutputStream(ByteArrayOutputStream output) {
      this.output = output;
    }

    @Override
    public long getPos() {
      return output.size();
    }

    @Override
    public void write(int b) throws IOException {
      output.write(b);
    }

    @Override
    public void write(byte[] b, int off, int len) throws IOException {
      output.write(b, off, len);
    }

    @Override
    public void close() throws IOException {
      output.close();
    }
  }
}
```

adam-christian-software replied:
Great! This looks to be handled. @snazy - We can resolve this, right?

dimas-b replied:
@flyrain : WDYT? blocker resolved?

pingtimeout reviewed (dismissed):
Thanks @snazy!  This is a well-designed PR that introduces valuable functionality for Iceberg table maintenance operations.  Apart from a couple suggestions, the PR LGTM.

I cannot think of any argument to block this PR if it reuses some classes defined in Project Nessie.  That being said, I am not opposed to pulling said classes in Polaris, if this suits the community better.  I recommend moving forward with this PR as-is and opening a Github issue so that the relevant classes are pulled in Polaris in a follow-up PR, for the sake of keeping this PR size manageable.  Any strong objection ?  cc @flyrain @dennishuo @singhpk234 @adam-christian-software

dimas-b reviewed (dismissed):
I support @pingtimeout suggestion to unblock this PR if there are no strong objections. Follow-up changes are certainly welcome (including in the areas that current use Nessie tools for tests).

CC: @dennishuo @flyrain @adam-christian-software @singhpk234

RussellSpitzer commented on storage/files/api/src/main/java/org/apache/polaris/storage/files/api/FileOperations.java:
```diff
@@ -0,0 +1,170 @@
+   * views. Rate-limiting on a single invocation may not be effective as expected.
+   *
+   * @param viewMetadataLocation Iceberg view-metadata location
+   * @param deduplicate if true, attempt to deduplicate files by their location, adding additional
```
Again i'm not sure why we need a deduplicate since I would think we can get the unique listing without any additional processing, I may be missing something here.

RussellSpitzer commented on storage/files/api/src/main/java/org/apache/polaris/storage/files/api/PurgeSpec.java:
```diff
@@ -0,0 +1,99 @@
+  PurgeSpec withPurgeIssuedCallback(Consumer<String> purgeIssuedCallback);
+
+  /**
+   * Optional rate-limit on the number of individual file-deletions per second.
```
number of individual files deleted per second?

file-deletions just seems like the spec also is allowing for individual file deletions outside of bulk deletions

RussellSpitzer commented on storage/files/impl/src/main/java/org/apache/polaris/storage/files/impl/FileOperationsImpl.java:
```diff
@@ -0,0 +1,472 @@
+    return fileSources.stream().flatMap(Function.identity());
+  }
+
+  static Predicate<String> deduplicator(boolean deduplicate) {
```
I mentioned this above, but I think we can skip the "deduplicator" logic all together here by using a bit more of the information in the Iceberg Table.

Starting with the oldest snapshot, add all manifests and entries, then for all newer snapshots add only manifests which have been marked as added and the entries within those which have been added. 

This only gets messed up if there is are other snapshots without parents (branches without a common snapshot, orphaned WAP - that one is probably not possible?) so we could handle duplicates for that, but it should only require-deduping at the manifest level I think. We only need to check added manifests in the snapshot and make sure they are deduped and then we can use the added/existing status within those manifests.

RussellSpitzer replied:
Not that any of this is important now, I just think it's something that could greatly reduce the amount of things required to deduplicate

RussellSpitzer commented on storage/files/impl/src/main/java/org/apache/polaris/storage/files/impl/FileOperationsImpl.java:
```diff
@@ -0,0 +1,472 @@
+          // Object stores do delete the files that exist, but a BulkDeletionFailureException is
+          // still being thrown.
+          // However, not all FileIO implementations behave the same way as some don't throw in the
+          // non-existent-case.
```
+1

RussellSpitzer reviewed (dismissed):
Overall looks pretty straight forward to me, I think there are some optimizations we can do for the streaming approach of listing files in an Iceberg table but not blockers for the moment. 

I do think the Nessie dependency in testing should probably be removed but that can happen after this PR is in. As I noted below I think either copying / reimplementing are both probably preferable to depending on another project. 

Also added a few notes since I think we currently aren't testing V3 tables (with puffin delete vector files) and it probably would be good to also test on tables with branches or WAP commits.

snazy commented:
There is always the opportunity to improve things later on. I do not think that there will be any new insights in the next 4/5 days, simply because there were no other proposals [since August 15](https://lists.apache.org/thread/9pgvhr9btfgzofbm6qhyfyqnk62hzp4m) or strong technical concerns.

pingtimeout commented:
@flyrain @RussellSpitzer I am working on the removal of the `object-storage-mock` and other test dependencies.  I should be able to propose something today or tomorrow.

Would it be possible to do that in a follow-up PR? Rationale: this one has already been increased by +1k LoC with the removal of Nessie code in the production scope.  And I think the testing scope would be trivial to review in isolation.

In other words: do you consider this dependency to Nessie libs for testing purposes as a blocker?

pingtimeout commented:
@RussellSpitzer I understand where you are coming from.  I just want to note that the size of this PR has already increased by 500 LoC since it was opened because the dependency to a Nessie library was considered as a blocker.  And I think it would be unusual to first block a PR because it depends on an external library, then request the code to be pulled into Polaris, and then block it again because it became too big.

I think I have seen multiple PRs in the project where contributions, however imperfect, are welcome, with the request that things are quickly fixed in follow-up PRs.  I am sure I can do some digging to find examples.  Isn't that a normal thing in OSS projects?

flyrain commented:
Echo this comment, https://github.com/apache/polaris/pull/3415#discussion_r2704149700, I think it's a good idea to test the code at scale, thanks @pingtimeout for bringing it up. Is there any benchmark for this PR? Can I reuse it or use it as a reference, so that we can compare apple to apple?  If not, I'd suggest to work together on a benchmark, so that it will benefits both PRs and any future improvement. WDYT?

RussellSpitzer commented:
As I mentioned before, it's just my opinion on how communities should operate, if you are committed to removing the Nessie dependencies then you already have my +1 on this PR.

I don't even have any positive feelings on keeping the current implementation, I would rather we just replace it entirely. For me it's just procedural issues in getting code into the project.

flyrain reviewed (changes requested):
Did another round of review. Left a few comments. Most of them are needed to be addressed, esp. https://github.com/apache/polaris/pull/3256#discussion_r2726474020. 

I'd suggest to publish benchmark as memory pressure may still be there, given the comment here https://github.com/apache/polaris/pull/3256#discussion_r2726474020. 

I'd also suggest to address my old comment from last year Dec 15, 2025, https://github.com/apache/polaris/pull/3256#pullrequestreview-3580578696. Addressing comments actively speeds up PR review process.

snazy reviewed (commented):
Pushed a couple of new commits to address the comments.

flyrain commented on storage/files/impl/src/main/resources/META-INF/beans.xml:
```diff
@@ -0,0 +1,24 @@
+<beans xmlns="https://jakarta.ee/xml/ns/jakartaee"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/beans_4_0.xsd">
+    <!-- File required by Weld (used for testing), not by Quarkus -->
```
We could always add it back once it is needed.

dimas-b commented on storage/files/impl/src/main/java/org/apache/polaris/storage/files/impl/StorageUri.java:
```diff
@@ -0,0 +1,288 @@
+ * <p>This type is <em>private</em> to the {@code polaris-storage-files-impl} module.
+ */
+final class StorageUri implements Comparable<StorageUri> {
+  public static final String SCHEME_FILE = "file";
```
The latest code does not have this constant.

snazy committed:
Add object-storage-mock test utility

This PR donates the polaris-object-storage-mock module to Polaris.

The module provides an embeddable object storage mock server for tests.
It exposes provider-facing HTTP surfaces for S3, GCS, ADLS Gen2, and STS, while keeping the backing bucket behavior controllable by tests. This lets Polaris tests exercise real SDK/client behavior without depending on external cloud services.

See tools/object-storage-mock/README.md for more detail on the module’s purpose, how it differs from a simple file-system or in-memory object store, and how it can be used.

This PR is intentionally scoped to the initial donation/import of the object-storage-mock into Polaris. Follow-up changes, cleanup, feature work, broader integration, or behavior changes should be handled in separate PRs.

snazy committed:
Donate Azurite and fake-gcs-server testcontainers

This PR donates reusable Testcontainers-based utilities for Azurite and fake-gcs-server to Polaris.

The new testcontainer modules provide local, containerized object-storage fixtures that can be used by Polaris tests without requiring external cloud services or credentials.
They are intended as an initial import/donation of the existing test infrastructure so Polaris can build follow-up integration work on top of a shared baseline.

This PR is intentionally scoped to the donation of these testcontainer utilities.
Cleanup, package renaming, additional documentation, broader test integration, or behavior changes should be handled in follow-up PRs.
